### PR TITLE
Add test case for short option handling

### LINKED ISSUE #1254: StringSlice() with UseShortOptionHandling causes duplicated entries, depending on the ordering of the flags
## my urfave/cli version is

_**v2.3.0**_

## Checklist

* [X] Are you running the latest v2 release? The list of releases is [here](https://github.com/urfave/cli/releases).
* [X] Did you check the manual for your release? The v2 manual is [here](https://github.com/urfave/cli/blob/master/docs/v2/manual.md)
* [X] Did you perform a search about this problem? Here's the [Github guide](https://help.github.com/en/github/managing-your-work-on-github/using-search-to-filter-issues-and-pull-requests) about searching.

## Dependency Management

- [X] My project is using go modules.

## Describe the bug

`StringSlice()` with `UseShortOptionHandling` causes duplicated entries, depending on the ordering of the flags.

## To reproduce

`main.go`:
```go
package main

import (
        "fmt"
        "os"

        "github.com/urfave/cli/v2"
)

func main() {
        app := cli.NewApp()
        app.Description = "minimal reproducer of https://github.com/AkihiroSuda/nerdctl/issues/108"
        app.UseShortOptionHandling = true
        app.Flags = []cli.Flag{
                &cli.BoolFlag{
                        Name: "i",
                },
                &cli.BoolFlag{
                        Name: "t",
                },
                &cli.StringSliceFlag{
                        Name: "net",
                },
        }
        app.Action = func(clicontext *cli.Context) error {
                fmt.Printf("clicontext.StringSlice(\"net\")=%+v\n", clicontext.StringSlice("net"))
                return nil
        }
        if err := app.Run(os.Args); err != nil {
                panic(err)
        }
}
```

`go.mod`:
```go
module example.com/reproduce-nerdctl-issue-108

go 1.16

require github.com/urfave/cli/v2 v2.3.0
```

Build: 
```console
$ go mod tidy
$ go build
```

## Observed behavior

```console
$ ./reproduce-nerdctl-issue-108 -it --net=foo
clicontext.StringSlice("net")=[foo]
$ ./reproduce-nerdctl-issue-108 --net=foo -it
clicontext.StringSlice("net")=[foo foo]
$ ./reproduce-nerdctl-issue-108 --net=foo -i -t
clicontext.StringSlice("net")=[foo]
```

The second command (`... -net=foo -it`) produces unexpectedly duplicated slice.

## Expected behavior

```console
$ ./reproduce-nerdctl-issue-108 -it --net=foo
clicontext.StringSlice("net")=[foo]
$ ./reproduce-nerdctl-issue-108 --net=foo -it
clicontext.StringSlice("net")=[foo]
$ ./reproduce-nerdctl-issue-108 --net=foo -i -t
clicontext.StringSlice("net")=[foo]
```

## Additional context

This bug needs to be fixed for nerdctl (Docker-compatible CLI for containerd): https://github.com/AkihiroSuda/nerdctl/issues/108

## Want to fix this yourself?
Not sure

## Run `go version` and paste its output here

```
go version go1.16.1 linux/amd64
```

## Run `go env` and paste its output here

```
GO111MODULE="on"
GOARCH="amd64"
GOBIN="/home/suda/gopath/bin"
GOCACHE="/home/suda/.cache/go-build"
GOENV="/home/suda/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/suda/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/suda/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.16.1"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/suda/reproduce-nerdctl-issue-108/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build2579704448=/tmp/go-build -gno-record-gcc-switches"
```

### PR DESCRIPTION
**Pull request by dearchap**
…erminated error

## What type of PR is this?

- bug

## What this PR does / why we need it:

The short options handling needs to restart from failed processing rather than from beginning of arguments
## Which issue(s) this PR fixes:

Fixes #1254

## Testing

go test -run=TestSliceShortOptionHandle ./...

## Release Notes

```release-note

```

### CODE REVIEW

**Comment by dearchap** [comment]
Bump

**Comment by dearchap** [comment]
Ping. Simple change and its just in test code . @coilysiren @rliebz

File: flag_test.go [hunk 1]
```diff
@@ -2050,3 +2050,34 @@ func TestTimestampFlagApply_WithDestination(t *testing.T) {
...
+// StringSlice() with UseShortOptionHandling causes duplicated entries, depending on the ordering of the flags
+func TestSliceShortOptionHandle(t *testing.T) {
+	_ = (&App{
+		Commands: []*Command{
+			{
+				Name:                   "foobar",
+				UseShortOptionHandling: true,
+				Action: func(ctx *Context) error {
+					if ctx.Bool("i") != true {
+						t.Errorf("bool i not set")
```
Comment (rliebz) [inline flag_test.go:2064]: Nit: `t.Error`

File: flag_test.go [hunk 2]
```diff
@@ -2050,3 +2050,34 @@ func TestTimestampFlagApply_WithDestination(t *testing.T) {
...
+					}
+					if ctx.Bool("t") != true {
+						t.Errorf("bool i not set")
```
Comment (rliebz) [inline flag_test.go:2067]: Nit: `t.Error`

File: flag_test.go [hunk 3]
```diff
@@ -2050,3 +2050,34 @@ func TestTimestampFlagApply_WithDestination(t *testing.T) {
 	expect(t, err, nil)
 	expect(t, *fl.Destination.timestamp, expectedResult)
 }
+
+// Test issue #1254
+// StringSlice() with UseShortOptionHandling causes duplicated entries, depending on the ordering of the flags
+func TestSliceShortOptionHandle(t *testing.T) {
+	_ = (&App{
```
Comment (rliebz) [inline flag_test.go:2057]: Nit: Can we add a quick nil error check here to make sure nothing unexpected is happening?

e.g.:

```go
err := (&App{
// ...

if err != nil {
  t.Fatal(err)
}

File: flag_test.go [hunk 4]
```diff
@@ -2050,3 +2050,34 @@ func TestTimestampFlagApply_WithDestination(t *testing.T) {
...
+		Commands: []*Command{
+			{
+				Name:                   "foobar",
+				UseShortOptionHandling: true,
+				Action: func(ctx *Context) error {
```
Comment (rliebz) [inline flag_test.go:2062]: One of the things that makes me nervous about only having assertions in a callback function is that the test passes if the callback function is never called. What I typically do in this kind of situation is add a boolean check:

```go
wasCalled := false

// ...

Action: func(ctx * Context) error {
  wasCalled = true
  // ...
},

// ...

// at the very end of the function
if !wasCalled {
  t.Error("missing call to action function")
}
```

**Review by rliebz — COMMENTED** [review]
Apologies for the delay, and thanks for the contribution!

Just wanted to double check, issue #1254 was already fixed in some other change set, and the purpose for this PR is to add a regression test?

File: flag_test.go (hunk 1 above)
Comment (dearchap) [reply flag_test.go:2064]: Done.

File: flag_test.go (hunk 2 above)
Comment (dearchap) [reply flag_test.go:2067]: Done.

File: flag_test.go (hunk 3 above)
Comment (dearchap) [reply flag_test.go:2057]: Done.

File: flag_test.go (hunk 4 above)
Comment (dearchap) [reply flag_test.go:2062]: Done.

**Review by meatballhat — APPROVED** [review]
🎉 Thank you, @dearchap!

**Pull request merged**